pandas/core/internals/managers.py

```diff
@@ -602,16 +602,27 @@ def comp(s, regex=False):
                 for b in rb:
                     m = masks[i][b.mgr_locs.indexer]
                     convert = i == src_len
-                    result = b._replace_coerce(
-                        mask=m,
-                        to_replace=s,
-                        value=d,
-                        inplace=inplace,
-                        convert=convert,
-                        regex=regex,
-                    )
                     if m.any() or convert:
-                        new_rb = _extend_blocks(result, new_rb)
+                        if m.ndim > 1:
```

Comment:
so you can use _block_shape to make all blocks conform to ndim=2, then you can simply separate this cleanly. this is too nested now

also would be ok to move parts of this to a helper function as this is getting hard to read.

further you might be able to discriminate the blocks on line 603 rather than here.

Reply:
Thanks for your review - I've tried using `_block_shape` from `pandas.core.internals.blocks import _block_shape`, but then I run into `AttributeError`s: 
```
>>> b
IntBlock: 4 dtype: int64
>>> b.values
array([1, 2, 3, 4])
>>> _block_shape(b, ndim=2)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/marco/pandas/pandas/core/internals/blocks.py", line 3053, in _block_shape
    values = values.reshape(tuple((1,) + shape))
AttributeError: 'IntBlock' object has no attribute 'reshape
```

sure, will work on the rest

Reply:
block_shape operates on an ndarray.

Reply:
Thanks for clarifying that. I've spent some time today thinking about this, but it's not clear to me how to
> use _block_shape to make all blocks conform to ndim=2
if `_block_shape` only operates on ndarrays. I can only see how to use it to ensure `m` is 2-dimensional.

If anyone following along can see how to do this, feel free to take over, it might currently be too advanced for me :) Else I'll keep working on it, it just might take a bit longer.

For now, I've split this part of the code into a helper function.